crates/typst/src/visualize/pattern.rs

```diff
@@ -0,0 +1,241 @@
+/// repeated in a grid-like fashion covering the entire area of the element
+/// being filled. You can also specify a spacing between the cells of the
+/// pattern, which is defined by a horizontal and vertical spacing. The spacing
+/// is the distance between the edges of adjacent cells of the pattern. The default
+/// spacing is zero.
+///
+/// # Examples
+///
+/// ```example
+/// #let pat = pattern((30pt, 30pt))[
+///   #place(top + left, line(start: (0%, 0%), end: (100%, 100%), stroke: 1pt))
+///   #place(top + left, line(start: (0%, 100%), end: (100%, 0%), stroke: 1pt))
+/// ]
+///
+/// #rect(fill: pat, width: 100%, height: 100%, stroke: 1pt)
+/// ```
+///
+/// Patterns are also supported on text, but only when setting the
+/// [relativeness]($pattern.relative) to either `{auto}` (the default value) or
+/// `{"parent"}`. To create word-by-word or glyph-by-glyph patterns, you can
+/// wrap the words or characters of your text in [boxes]($box) manually or
+/// through a [show rule]($styling/#show-rules).
+///
+/// ```example
+/// #let pat = pattern(
+///   (30pt, 30pt),
+///   relative: "parent",
+///   square(size: 30pt, fill: gradient.conic(..color.map.rainbow))
+/// );
+///  #set text(fill: pat)
+///  #lorem(10)
+/// ```
+///
+/// You can also space the elements further or closer apart using the
+/// [`spacing`]($pattern.spacing) feature of the pattern. If the spacing
+/// is lower than the size of the pattern, the pattern will overlap.
+/// If it is higher, the pattern will have gaps of the same color as the
+/// background of the pattern.
+///
+/// ```example
+/// #let pat = pattern(
+///   (30pt, 30pt),
+///   spacing: (10pt, 10pt),
+///   relative: "parent",
+///   square(size: 30pt, fill: gradient.conic(..color.map.rainbow))
+/// );
+///  #rect(width: 100%, height: 100%, fill: pat)
+/// ```
+///
+/// # RelativeToness
+/// The location of the starting point of the pattern is dependant on the
+/// dimensions of a container. This container can either be the shape they
+/// are painted on, or the closest surrounding container. This is controlled by
+/// the `relative` argument of a pattern constructor. By default, patterns are
+/// relative to the shape they are painted on, unless the pattern is applied on
+/// text, in which case they are relative to the closest ancestor container.
+///
+/// Typst determines the ancestor container as follows:
+/// - For shapes that are placed at the root/top level of the document, the
+///   closest ancestor is the page itself.
+/// - For other shapes, the ancestor is the innermost [`block`]($block) or
+///   [`box`]($box) that contains the shape. This includes the boxes and blocks
+///   that are implicitly created by show rules and elements. For example, a
+///   [`rotate`]($rotate) will not affect the parent of a gradient, but a
+///   [`grid`]($grid) will.
+#[ty(scope)]
+#[derive(Debug, Clone, PartialEq, Hash)]
+pub struct Pattern {
```

Comment:
This type will significantly increase the size of `Value` again. perhaps the whole data should be in a `struct Repr` and then pattern is `Pattern(Arc<Prehashed<Repr>>)`.

Reply:
Yes you're right I hadn't considered it, changed!